### PR TITLE
feat: add kernels for Tensor Parallel

### PR DESCRIPTION
**Pull request by Chamberlain0w0**
添加 Tensor Parallel 所需要的 kernel/function/module：

全局添加的包括，Tensor 与 functional 里面可能会新增了对应的接口：
1. elementwise: exp/log/equals 以及各种逻辑运算
2. Gather：kernel 起名为 IndexGather，对齐 torch.gather
3. Concat：对齐 torch.cat 
4. 通信：AllGather 以及 Reduce Scatter
5. ~仅用于 TP 的 VocabParallelCrossEntropyBackwardKernel~
6. ~添加 Tensor::Clone()~

全局修改的包括（主要是 cuda kernel）：
1. elementwise.cu：
    a. 因为 TP 中构造 mask 涉及 int64 的 index 的运算，所以在 unary/binary 前向反向中给 dispatch 加了 int64 的分支（Binary 反向没加，因为 atomicadd 不支持 int64，同时构造 mask 的操作里的 tensor 都是 requires_grad=False，没用上这个反向；留了个 fixme 的注释）
    b. 个别用到 `decltype(x){0.f};` 的地方改了下写法，消除了关于没法安全从 float 转为 double 的 warning（因为 x 可能是 double）
2. embedding.cu 里面添加关于 token id 超越词表的判断（TP 里要对 vocab 切分）
3. reduction.cu 里面也是改了下原先 cub::Sum 等接口（deprecated），为了消除 warning
4. transform.cu 里的 MaskedFill，原先是只支持 mask 形状和 input 形状的右对齐，现在是加了个动态判断来支持左对齐，对应的 python 语义类似于 `input[mask, :] = value`

~TP 所需要的全部位于 tensor_parallel.h/.cc 里面：~
1. ~TensorParallelGroup 定义，后续可能会改并考虑和 DDP 的通信组的融合~
2. ~一些包装了通信的函数，比如 CopyToTPRegionFunc 等，内部对应的就是一个对应语义的 autograd::Function 的 Apply 调用~
3. ~ColumnParallelLinear/RowParallelLinear/VocabParallelEmbedding，基本对齐 Megatron-LM 实现方式~
4. ~VocabParallelCrossEntropy 的 Function 和 Module~
    a. ~为了尽可能对齐 Megatron 的逻辑也加入了 label smoothing，以便后续要跟 megatron 比精度~
    b. ~function 的前向是小算子拼接，反向是一个大 kernel，前向输出为 loss tensor~
    c. ~module 里在调用 function 的 apply 的基础上又做了个全局的 mean，输出 scalar loss~

### CODE REVIEW

**Comment by Chamberlain0w0** [comment]
更新：仅保留新增通用 kernel/function，TP 相关建设在本 PR 都删掉了，等待后续合入

**Review by kilinchange — CHANGES_REQUESTED** [review]

File: infini_train/src/tensor.cc [hunk 1]
```diff
@@ -224,11 +225,81 @@ Tensor Tensor::To(DataType dtype) {
     return new_tensor;
 }
 
+Tensor Tensor::Clone() const {
```
[the remark this answers is not in the document]
Comment (Chamberlain0w0) [reply infini_train/src/tensor.cc:228]: 仔细想了想我这边其实并不需要，当时是想着构图的时候不能直接把 input 的指针当 output 返回所以整了这么一出，但实际上在构图是新创一个指针指向原来的对象似乎就够了。我先把这块删了吧

**Review by kilinchange — APPROVED** [review]

**Review by Ziminli — APPROVED** [review]
